homeassistant/components/alarm_control_panel/spc.py

```diff
@@ -4,71 +4,63 @@
 """
-import asyncio
 import logging
 
 import homeassistant.components.alarm_control_panel as alarm
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
 from homeassistant.components.spc import (
-    ATTR_DISCOVER_AREAS, DATA_API, DATA_REGISTRY, SpcWebGateway)
+    ATTR_DISCOVER_AREAS, DATA_API, SIGNAL_UPDATE_ALARM)
 from homeassistant.const import (
-    STATE_ALARM_ARMED_AWAY, STATE_ALARM_ARMED_HOME, STATE_ALARM_DISARMED,
-    STATE_UNKNOWN)
+    STATE_ALARM_ARMED_AWAY, STATE_ALARM_ARMED_HOME, STATE_ALARM_ARMED_NIGHT,
+    STATE_ALARM_DISARMED, STATE_ALARM_TRIGGERED)
 
 _LOGGER = logging.getLogger(__name__)
 
-SPC_AREA_MODE_TO_STATE = {
-    '0': STATE_ALARM_DISARMED,
-    '1': STATE_ALARM_ARMED_HOME,
-    '3': STATE_ALARM_ARMED_AWAY,
-}
 
-
-def _get_alarm_state(spc_mode):
+def _get_alarm_state(area):
     """Get the alarm state."""
-    return SPC_AREA_MODE_TO_STATE.get(spc_mode, STATE_UNKNOWN)
+    from pyspcwebgw.const import AreaMode
+
+    if area.verified_alarm:
+        return STATE_ALARM_TRIGGERED
+
+    mode_to_state = {
+        AreaMode.UNSET: STATE_ALARM_DISARMED,
+        AreaMode.PART_SET_A: STATE_ALARM_ARMED_HOME,
+        AreaMode.PART_SET_B: STATE_ALARM_ARMED_NIGHT,
+        AreaMode.FULL_SET: STATE_ALARM_ARMED_AWAY,
+    }
+    return mode_to_state.get(area.mode, None)
 
 
-@asyncio.coroutine
-def async_setup_platform(hass, config, async_add_entities,
-                         discovery_info=None):
+async def async_setup_platform(hass, config, async_add_entities,
+                               discovery_info=None):
     """Set up the SPC alarm control panel platform."""
     if (discovery_info is None or
             discovery_info[ATTR_DISCOVER_AREAS] is None):
         return
 
-    api = hass.data[DATA_API]
-    devices = [SpcAlarm(api, area)
-               for area in discovery_info[ATTR_DISCOVER_AREAS]]
-
-    async_add_entities(devices)
+    async_add_entities([SpcAlarm(area=area, api=hass.data[DATA_API])
+                        for area in discovery_info[ATTR_DISCOVER_AREAS]])
 
 
 class SpcAlarm(alarm.AlarmControlPanel):
     """Representation of the SPC alarm panel."""
 
-    def __init__(self, api, area):
+    def __init__(self, area, api):
         """Initialize the SPC alarm panel."""
-        self._area_id = area['id']
-        self._name = area['name']
-        self._state = _get_alarm_state(area['mode'])
-        if self._state == STATE_ALARM_DISARMED:
-            self._changed_by = area.get('last_unset_user_name', 'unknown')
-        else:
-            self._changed_by = area.get('last_set_user_name', 'unknown')
+        self._area = area
         self._api = api
 
-    @asyncio.coroutine
-    def async_added_to_hass(self):
+    async def async_added_to_hass(self):
         """Call for adding new entities."""
-        self.hass.data[DATA_REGISTRY].register_alarm_device(
-            self._area_id, self)
+        async def async_update_callback():
```

Comment:
Make this a callback instead of a coroutine by decorating the function with `@callback` imported from core.py.